freqtrade/freqtradebot.py

```diff
@@ -845,44 +845,60 @@ def handle_timedout_limit_buy(self, trade: Trade, order: Dict) -> bool:
         :return: True if order was fully cancelled
         """
         reason = "cancelled due to timeout"
-        if order['status'] != 'canceled':
-            corder = self.exchange.cancel_order(trade.open_order_id, trade.pair)
-        else:
-            # Order was cancelled already, so we can reuse the existing dict
-            corder = order
-            reason = "canceled on Exchange"
 
-        if corder.get('remaining', order['remaining']) == order['amount']:
-            # if trade is not partially completed, just delete the trade
-            self.handle_buy_order_full_cancel(trade, reason)
-            return True
+        # running get_signal on historical data fetched
+        (buy, sell) = self.strategy.get_signal(
+            trade.pair, self.strategy.ticker_interval,
+            self.dataprovider.ohlcv(trade.pair, self.strategy.ticker_interval))
 
-        # if trade is partially complete, edit the stake details for the trade
-        # and close the order
-        # cancel_order may not contain the full order dict, so we need to fallback
-        # to the order dict aquired before cancelling.
-        # we need to fall back to the values from order if corder does not contain these keys.
-        trade.amount = order['amount'] - corder.get('remaining', order['remaining'])
-        trade.stake_amount = trade.amount * trade.open_rate
-        # verify if fees were taken from amount to avoid problems during selling
-        try:
-            new_amount = self.get_real_amount(trade, corder if 'fee' in corder else order,
-                                              trade.amount)
-            if not isclose(order['amount'], new_amount, abs_tol=constants.MATH_CLOSE_PREC):
-                trade.amount = new_amount
-                # Fee was applied, so set to 0
-                trade.fee_open = 0
-                trade.recalc_open_trade_price()
-        except DependencyException as e:
-            logger.warning("Could not update trade amount: %s", e)
-
-        trade.open_order_id = None
-        logger.info('Partial buy order timeout for %s.', trade)
-        self.rpc.send_msg({
-            'type': RPCMessageType.STATUS_NOTIFICATION,
-            'status': f'Remaining buy order for {trade.pair} cancelled due to timeout'
-        })
-        return False
+        # get config for bid strategy
+        config_bid_strategy = self.config.get('bid_strategy', {})
+
+        # proceed to cancel buy order by timeout if timeout_even_if_buy_signal_valid
+        # is true (original behaviour) -OR-
+        # cancel buy order only if buying condition is no longer valid OR if there's
+        # a sell signal present
+        if config_bid_strategy.get('timeout_even_if_buy_signal_valid', True) or (not buy or sell):
```

Comment:
personally, i'd move the "decision logic" out of this method and into `check_handle_timedout()`.

That'll give us 2 advantages:
* Implementing this for sell-side is simpler
* the decision is taken before this method - so this methods' purpose is to apply the decision, not to decide anything.

Reply:
Makes sense, I'll make a commit even if we're not there yet in terms of how this should finally look like, just to bring it closer to reality.